crates/bevy_math/src/cubic_splines.rs

```diff
@@ -279,9 +290,269 @@ impl<P: Point> CubicGenerator<P> for CubicBSpline<P> {
+        }
+
+        // Check the weights vector length
+        if weights.len() != control_points_len {
+            return Err(CubicNurbsError::WeightsVectorMismatch {
+                expected: control_points_len,
+                provided: weights.len(),
+            });
+        }
+
+        weights = Self::normalize_weights(weights);
+
+        control_points
+            .iter_mut()
+            .zip(weights)
+            .for_each(|(p, w)| *p = *p * w);
+
+        Ok(Self {
+            control_points,
+            knot_vector,
+        })
+    }
+
+    /// Generates a uniform knot vector that will generate the same curve as [`CubicBSpline`].
+    ///
+    /// "Uniform" means that the difference between two knot values next to each other is the same
+    /// through the entire knot vector.
+    ///
+    /// Will return `None` if there are less than 4 control points
+    pub fn uniform_knot_vector(control_points: usize) -> Option<Vec<f32>> {
+        if control_points < 4 {
+            return None;
+        }
+        Some(
+            (0..Self::knot_vector_length(control_points))
+                .map(|v| v as f32)
+                .collect(),
+        )
+    }
+
+    /// Generates an open uniform knot vector, which makes the ends of the curve pass through the
+    /// start and end points.
+    ///
+    /// The knot vector will have a knot with multiplicity of 4 at the end and start and elements
+    /// in the middle will have a difference of 1. "Multiplicity" means taht there are N
+    /// consecutive elements that have the same value.
+    ///
+    /// Will return `None` if there are less than 4 control points
+    pub fn open_uniform_knot_vector(control_points: usize) -> Option<Vec<f32>> {
+        if control_points < 4 {
+            return None;
+        }
+        let last_knots_value = control_points - 3;
+        Some(
+            std::iter::repeat(0.0)
+                .take(4)
+                .chain((1..last_knots_value).map(|v| v as f32))
+                .chain(std::iter::repeat(last_knots_value as f32).take(4))
+                .collect(),
+        )
+    }
+
+    #[inline(always)]
+    const fn knot_vector_length(control_points_len: usize) -> usize {
+        control_points_len + 4
+    }
+
+    /// Based on <https://xiaoxingchen.github.io/2020/03/02/bspline_in_so3/general_matrix_representation_for_bsplines.pdf>
+    fn generate_matrix(knot_vector_segment: &[f32; 8]) -> [[f32; 4]; 4] {
+        let t = knot_vector_segment;
+        let m00 = (t[4] - t[3]).powi(2) / ((t[4] - t[2]) * (t[4] - t[1]));
+        let m02 = (t[3] - t[2]).powi(2) / ((t[5] - t[2]) * (t[4] - t[2]));
+        let m12 = (3.0 * (t[4] - t[3]) * (t[3] - t[2])) / ((t[5] - t[2]) * (t[4] - t[2]));
+        let m22 = 3.0 * (t[4] - t[3]).powi(2) / ((t[5] - t[2]) * (t[4] - t[2]));
+        let m33 = (t[4] - t[3]).powi(2) / ((t[3 + 3] - t[3]) * (t[5] - t[3]));
+        let m32 = -m22 / 3.0 - m33 - (t[4] - t[3]).powi(2) / ((t[5] - t[3]) * (t[5] - t[2]));
```

Comment:
These calculations are taken straight from the paper linked in the comment above. I think it works as intended for those cases.

Reply:
Can we add tests to validate this?

Reply:
These calculations are not used when multiple values in the knot vector are the same because the segment length becomes zero. We should test that zero length segments are handled properly, but we don't need to worry about this matrix.